File: OrangeJuice.Server.Core/Services/AzureClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

using Microsoft.WindowsAzure.Storage.Blob;
using Microsoft.WindowsAzure.Storage.Table;
using OrangeJuice.Server.Configuration;

namespace OrangeJuice.Server.Services
{
	public sealed class AzureClient : IAzureClient
	{
		private static readonly TimeSpan Year = TimeSpan.FromDays(365);

		private readonly IAzureContainerClient _containerClient;
		private readonly IBlobClient _blobClient;

		public AzureClient(IAzureContainerClient containerClient, IBlobClient blobClient)
		{
			_blobClient = blobClient;
			_containerClient = containerClient;
		}

		public async Task<string> GetBlobFromContainer(string containerName, string fileName)
		{
			ICloudBlob blob = await _containerClient.GetBlockReference(containerName, fileName);

			bool exists = await blob.ExistsAsync();
			if (!exists)
				return null;

			return await _blobClient.Read(blob);
		}

		public async Task<string[]> GetBlobsFromContainer(string containerName)
		{
			var container = await _containerClient.GetContainer(containerName);
			var tasks = container.ListBlobs()
								 .Cast<CloudBlockBlob>()
								 .Select(b => b.DownloadTextAsync());
			return await Task.WhenAll(tasks);
		}

		public async Task PutBlobToContainer(string containerName, string fileName, string content)
		{
			ICloudBlob blob = await _containerClient.GetBlockReference(containerName, fileName);
			blob.Properties.CacheControl = CreateCacheControl(Year);
			await _blobClient.Write(blob, content);
		}

		public async Task<Uri> GetBlobUrl(string containerName, string fileName)
		{
			CloudBlob blob = await _containerClient.GetBlobReference(containerName, fileName);
			bool exists = await blob.ExistsAsync();
			return exists ? blob.Uri : null;
		}

		public async Task<T[]> GetEntitiesFromTable<T>(string tableName)
			where T : ITableEntity, new()
		{
			var table = await _containerClient.GetTableReference(tableName);
			var query = new TableQuery<T>();
			return table.ExecuteQuery(query).ToArray();
		}

		public async Task<IList<TableResult>> PutEntitiesToTable<T>(string tableName, IEnumerable<T> entities)
			where T : ITableEntity
		{
			var table = await _containerClient.GetTableReference(tableName);

			var batch = new TableBatchOperation();
			foreach (var e in entities)
			{
				batch.Insert(e);
			}

			return await table.ExecuteBatchAsync(batch);
		}

		private static string CreateCacheControl(TimeSpan timeSpan)
		{
			return String.Format("public, max-age={0}", timeSpan.TotalMilliseconds);
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Services/AzureClient.fs
namespace OrangeJuice.Server.FSharp.Services

open System
open System.Collections.Generic
open System.Threading.Tasks

open Microsoft.WindowsAzure.Storage.Blob
open Microsoft.WindowsAzure.Storage.Table;

open OrangeJuice.Server.Services

type AzureClient(blobClient : IBlobClient, containerClient : IAzureContainerClient) =
    static let Year = TimeSpan.FromDays(365.0)

    static let CreateCacheControl = fun (timeSpan : TimeSpan) -> sprintf "public, max-age=%f" timeSpan.TotalMilliseconds

    interface IAzureClient with
        member this.GetBlobFromContainer(containerName : string, fileName : string) : Task<string> =
            let task = async {
                let! blob = containerClient.GetBlockReference(containerName, fileName) |> Async.AwaitTask
                let! exists = blob.ExistsAsync() |> Async.AwaitTask
                return!
                    match exists with
                        | true -> blobClient.Read(blob) |> Async.AwaitTask
                        | false -> Task.FromResult(null) |> Async.AwaitTask
            }
            task |> Async.StartAsTask

        member this.GetBlobsFromContainer(containerName : string) : Task<string[]> =
            let task = async {
                let! container = containerClient.GetContainer(containerName) |> Async.AwaitTask
                return! container.ListBlobs()
                        |> Seq.cast<CloudBlockBlob>
                        |> Seq.map (fun b -> b.DownloadTextAsync() |> Async.AwaitTask)
                        |> Async.Parallel
            }
            task |> Async.StartAsTask

        member this.PutBlobToContainer(containerName : string, fileName : string, content : string) : Task =
            let task = async {
                let! blob = containerClient.GetBlockReference(containerName, fileName) |> Async.AwaitTask
                blob.Properties.CacheControl <- CreateCacheControl(Year)
                return blobClient.Write(blob, content)
            }
            task |> Async.StartAsTask :> Task

        member this.GetBlobUrl(containerName : string, fileName : string) : Task<Uri> =
            let task = async {
                let! blob = containerClient.GetBlobReference(containerName, fileName) |> Async.AwaitTask
                let! exists = blob.ExistsAsync() |> Async.AwaitTask
                return
                    match exists with
                        | true -> blob.Uri
                        | false -> null
            }
            task |> Async.StartAsTask

        member this.GetEntitiesFromTable<'T when 'T :> ITableEntity  and 'T : (new : unit -> 'T)>(tableName : string) : Task<'T[]> =
            let task = async {
                let! table = containerClient.GetTableReference(tableName) |> Async.AwaitTask
                let query = new TableQuery<'T>();
                return table.ExecuteQuery(query)
                       |> Array.ofSeq
            }
            task |> Async.StartAsTask

        member this.PutEntitiesToTable<'T when 'T :> ITableEntity>(tableName : string, entities : seq<'T>) : Task<IList<TableResult>> =
            let task = async {
                let! table = containerClient.GetTableReference(tableName) |> Async.AwaitTask
                let batch = new TableBatchOperation()
                for e in entities do
                    batch.Insert(e)
                return! table.ExecuteBatchAsync(batch) |> Async.AwaitTask
            }
            task |> Async.StartAsTask